climada_petals/hazard/rf_glofas/transform_ops.py

```diff
@@ -0,0 +1,809 @@
+"""
+This file is part of CLIMADA.
+
+Copyright (C) 2017 ETH Zurich, CLIMADA contributors listed in AUTHORS.
+
+CLIMADA is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Transformations for dantro data pipeline
+"""
+
+import logging
+import re
+from pathlib import Path
+from typing import Optional, Union, List, Mapping, Any, Iterable, Tuple, Callable
+from collections import deque
+from copy import deepcopy
+
+import xarray as xr
+import numpy as np
+import pandas as pd
+import geopandas as gpd
+from scipy.stats import gumbel_r
+import xesmf as xe
+from numba import guvectorize
+
+from climada.util.constants import SYSTEM_DIR
+from climada.util.coordinates import get_country_geometries, country_to_iso
+from .cds_glofas_downloader import glofas_request
+
+LOGGER = logging.getLogger(__name__)
+
+CDS_DOWNLOAD_DIR = Path(SYSTEM_DIR, "cds-download")
+
+
+def sel_lon_lat_slice(target: xr.DataArray, source: xr.DataArray) -> xr.DataArray:
+    """Select a lon/lat slice from 'target' using coordinates of 'source'"""
+    lon = source["longitude"][[0, -1]]
+    lat = source["latitude"][[0, -1]]
+    return target.sel(longitude=slice(*lon), latitude=slice(*lat))
+
+
+def rp_comp(
+    sample: np.ndarray,
+    loc: np.ndarray,
+    scale: np.ndarray,
+    max_rp: Optional[float] = np.inf,
+):
+    """Compute the return period from a right-handed Gumbel distribution
+
+    All parameters can be arrays, in which case numpy broadcasting rules apply.
+
+    The return period of a sample :math:`x` from an extreme value distribution is
+    defined as :math:`(1 - \\mathrm{cdf}(x))^{-1}`, where :math:`\\mathrm{cdf}` is the
+    cumulative distribution function of said distribution.
+
+    Parameters
+    ----------
+    sample : array
+        Samples for which to compute the return period
+    loc : array
+        Loc parameter of the Gumbel distribution
+    scale : array
+        Scale parameter of the distribution
+    max_rp : float, optional
+        The maximum value of return periods. This avoids returning infinite values.
+        Defaults to ``np.inf`` (no maximum).
+
+    Returns
+    -------
+    np.ndarray
+        The return period(s) for the input parameters
+    """
+    return np.minimum(1.0 / (1.0 - gumbel_r.cdf(sample, loc=loc, scale=scale)), max_rp)
```

Comment:
Very good point! But wouldn't it be simpler (and safe because of floating point arithmetics) to *always* add the spacing?
```suggestion
    cdf = gumbel_r.cdf(sample, loc=loc, scale=scale)
    return np.fmin(1.0 / (1.0 - cdf + np.spacing(1.0)), max_rp)
```

Reply:
In cases where `cdf == 1.0`, my version will return `np.inf`, and your version will return a very large float. In my opinion, the case `cdf == 1.0` is in a different category of what can happen, so I would also want the code to be able to produce `np.inf` as a return value in this case. If users don't like `np.inf`, they are free to set `max_rp` to anything that they prefer. But if users want to have `np.inf`, e.g. for doing sanity checks, they should be able to get this value by setting `max_rp` to `np.inf` (which happens to be the default, actually).